### PR TITLE
Added JSON file account service

### PR DESCRIPTION
**Pull request by siepkes**
This PR adds an account service implementation which uses a local JSON file. The acount service checks the file at fixed intervals for changes. It's primarily useful for small deployments with little account changes and development setups.

It's functionally pretty complete. The unit test still need to be written though. However before I take it any further I was looking for some feedback from you guys. Are there any obvious things I missed? Are there any things (edge cases, etc.) I should be aware with the account service?

### CODE REVIEW

File: ambry-account/src/main/java/com/github/ambry/account/JsonAccountService.java [hunk 1]
```diff
@@ -0,0 +1,291 @@
...
+        newAccountNames.add(account.getName());
+        newAccountsList.add(account);
+      }
+
+      // Remove accounts from the account maps which don't exist anymore.
+      accountsById.keySet().retainAll(newAccountIds);
+      accountsByName.keySet().retainAll(newAccountNames);
+      accounts = Collections.unmodifiableCollection(newAccountsList);
+
+      notifyAccountUpdateConsumers(updatedAccounts);
```
[the remark this answers is not in the document]
Comment (siepkes) [reply ambry-account/src/main/java/com/github/ambry/account/JsonAccountService.java]: Fixed!

File: ambry-account/src/main/java/com/github/ambry/account/JsonAccountService.java [hunk 2]
```diff
@@ -0,0 +1,291 @@
...
+   * Updates the {@link #accountsById} based on the collection of supplied accountsById.
+   *
+   * Accounts not present in the map are added but present in the supplied collection are added.
+   * Accounts present in the map but not in the collection are evicted.
+   *
+   * @param newAccounts
+   */
+  private void updateAccountsMap(Set<Account> newAccounts) {
+    try {
+      accountMapLock.tryLock();
```
[the remark this answers is not in the document]
Comment (siepkes) [reply ambry-account/src/main/java/com/github/ambry/account/JsonAccountService.java]: Good catch. It's actually even worse: `tryLock()` returns a boolean which you need to check ;-) I needed `lock()`. I've corrected both.

File: ambry-account/src/main/java/com/github/ambry/account/JsonAccountService.java [hunk 3]
```diff
@@ -0,0 +1,291 @@
...
+   * @return The collection of {@link Account}s parsed from the given json file.
+   *
+   * @throws IOException Thrown when we can't open the JSON file with account data.
+   * @throws JSONException Thrown when we are unable to correctly parse the JSON account data in the JSON acocunt file.
+   * @throws IllegalArgumentException Thrown when the JSON file contains invalid data for Ambry. For example duplicate account ID's.
+   */
+  private Set<Account> getAccountsFromJsonFile() throws IOException, JSONException {
+    JSONArray accountArray = new JSONArray(Utils.readStringFromFile(accountFile.toString()));
+    // We'll only be iterating over the result in order so we use a linked list.
+    Set<Account> accounts = new HashSet<>();
```
[the remark this answers is not in the document]
Comment (siepkes) [reply ambry-account/src/main/java/com/github/ambry/account/JsonAccountService.java]: Modified it. No good reason anymore; Was a left over from a account notifier implementation where a set was useful.

File: ambry-account/src/main/java/com/github/ambry/account/JsonAccountService.java [hunk 4]
```diff
@@ -0,0 +1,291 @@
...
+              || accountFilePreviousModified.compareTo(accountFileLastModified) < 0) {
+            logger.debug("JSON Account file has changed or has never been loaded before. "
+                    + "Previous last know modified time: {}, new last known modified time: {}", accountFilePreviousModified,
+                accountFileLastModified);
+
+            updateAccountsFromFile();
+            accountFilePreviousModified = accountFileLastModified;
+          }
+
+          success = true;
```
[the remark this answers is not in the document]
Comment (siepkes) [reply ambry-account/src/main/java/com/github/ambry/account/JsonAccountService.java]: I didn't really like the use of this variable either (instead doing it in the `catch`, etc.) but I thought it was convention (I modeled this after the `HelixAccountService`). However looking at it again I realize now I misinterperted it. It's simply because the `helixStore.update` returns a boolean. I've modified it, let me know if this is more in line with your expectations!

**Comment by siepkes** [comment]
@vgkholla Thanks for taking the time to take a look at it! I've addressed the issues you've raised. I'll start working on adding some tests for the service.

**Comment by siepkes** [comment]
Currently working on tests. Will let you guys know when I think it's ready for another round of reviews.

**Comment by siepkes** [comment]
@vgkholla I've added a bunch of tests I thought would be meaningful. I think it's pretty complete overall. Let me know if you guys would like to see some additional tests, modifications, etc.

**Comment by vgkholla** [comment]
@siepkes thanks - will take a look (might take a while). 
I have a question reg. the implementation. Have you seen the impl of `HelixAccountService`? The data that that impl gets is actually a JSON from helix. The only way that impl differs from the one you are proposing is the location of the JSON (in helix v/s locally). Could you see if the impls can use common code once the JSON is obtained (processing the JSON, putting it in memory etc)? If its possible, i'd prefer to extract the core of the well tested code in `HelixAccountService` and get both this impl and `HelixAccountService` to use that core code

**Comment by siepkes** [comment]
@vgkholla I'll take a crack at extracting the common logic!

**Comment by vgkholla** [comment]
Thanks! I hope to review this soon - things are a little busy ATM !

File: ambry-account/src/test/java/com/github/ambry/account/JsonAccountServiceTest.java [hunk 5]
```diff
@@ -0,0 +1,264 @@
...
+    accountService.processAccountJsonFile();
+
+    assertEquals("Did not receive expected account added notifications.", notificationsReceived, addedAccounts);
+  }
+
+  /**
+   * Test if we receive notifications for modified accounts.
+   */
+  @Test
+  public void testdAccountModifyNotification() throws Exception {
```
Comment (cgtz) [inline ambry-account/src/test/java/com/github/ambry/account/JsonAccountServiceTest.java]: extra d in method name

File: ambry-account/src/main/java/com/github/ambry/account/JsonAccountService.java [hunk 6]
```diff
@@ -0,0 +1,235 @@
...
+
+  /**
+   * Starts a thread which checks (at fixed intervals) if the JSON account file has changed. If it has it triggers
+   * and update of the accounts.
+   */
+  private void watchJsonFile() {
+    watchAccountFileThread = new Thread(() -> {
+      logger.info("Starting JSON account file monitor. File check interval is {}ms", fileWwatchIntervalInMs);
+
+      while (!Thread.interrupted()) {
```
Comment (cgtz) [inline ambry-account/src/main/java/com/github/ambry/account/JsonAccountService.java]: Have you looked into using WatchService to react to file changes instead of using a polling interval: https://docs.oracle.com/javase/tutorial/essential/io/notification.html

File: ambry-account/src/main/java/com/github/ambry/account/JsonAccountService.java [hunk 7]
```diff
@@ -0,0 +1,235 @@
...
+  }
+
+  /** Guards against calling {@link #init} multiple times.*/
+  private final AtomicBoolean initialized = new AtomicBoolean();
+
+  /**
+   * Initializes this instance. Does all work which requires handing out references to this instance itself which can't
+   * be done during normal instance construction without letting a reference to an incomplete constructed instance escape.
+   */
+  void init() {
```
Comment (cgtz) [inline ambry-account/src/main/java/com/github/ambry/account/JsonAccountService.java]: In HelixAccountService we do similar initialization (with lambdas and passing references to the instance under construction) within the constructor. Is there an issue that arises when doing that?

Comment (cgtz) [reply ambry-account/src/main/java/com/github/ambry/account/JsonAccountService.java]: Ah, I found this post: https://stackoverflow.com/questions/2513597/what-is-an-incompletely-constructed-object

That is an interesting case. I guess it relates to the lambda's ability to see the final fields set in the constructor.

File: ambry-account/src/main/java/com/github/ambry/account/JsonAccountService.java (hunk 6 above)
Comment (cgtz) [reply ambry-account/src/main/java/com/github/ambry/account/JsonAccountService.java]: Ah, it actually seems like this method is not used, since it is handled by the updater `ScheduledExecutorService`, so this method could probably be removed

File: ambry-account/src/main/java/com/github/ambry/account/AbstractAccountService.java [hunk 8]
```diff
@@ -0,0 +1,331 @@
...
+
+
+/**
+ * Class which provides a basic implementation an {@link AccountService}. The handling of accounts is mostly the same
+ * for all the implementations. The only thing the implementations (usually) differ in is what the source is of the
+ * accounts (Zookeeper / Helix, local JSON file, etcd, Consul, etc).
+ */
+abstract class AbstractAccountService implements AccountService {
+
+	private static final Logger logger = LoggerFactory.getLogger(AbstractAccountService.class);
```
[the remark this answers is not in the document]
Comment (siepkes) [reply ambry-account/src/main/java/com/github/ambry/account/AbstractAccountService.java]: Done. I also checked all the other files for style.

File: ambry-account/src/test/java/com/github/ambry/account/JsonAccountServiceTest.java (hunk 5 above)
Comment (siepkes) [reply ambry-account/src/test/java/com/github/ambry/account/JsonAccountServiceTest.java]: Fixed!

File: ambry-account/src/main/java/com/github/ambry/account/JsonAccountService.java (hunk 6 above)
Comment (siepkes) [reply ambry-account/src/main/java/com/github/ambry/account/JsonAccountService.java]: I did consider using the `WatchService`. I can understand the the question so I have added the following comment to the class:

> This implementation uses a thread to check for changes in the JSON file at scheduled intervals. While using a {@link java.nio.file.WatchService} might seem more intuitive the docs of the WatchService specify that the characteristics of the WatchService are highly platform (Linux, Windows, Solaris, etc.) dependent. Going as far as specifying events may under certain conditions not be fired. Therefor having a fallback which at intervals checks if the JSON file has changed will probably always be needed to ensure no changes are missed even if a WatchService would be added.

Adding a `WatchService` to the implementation could be used to improve the change detection. But I don't think that one can do with the regular intervals checks to ensure all changes are noticed.

File: ambry-account/src/main/java/com/github/ambry/account/JsonAccountService.java (hunk 7 above)
Comment (siepkes) [reply ambry-account/src/main/java/com/github/ambry/account/JsonAccountService.java]: Yes Lambda's added a new risk for letting the `this` reference "escape". It usually occurred when someone registered a callback from a constructor. The "no escaping" rule has been part of Java Concurrency in practice for a long time.

**Comment by siepkes** [comment]
@cgtz No sweat! I know how these things go. Thanks for the review!

**Comment by justinlin-linkedin** [comment]
@siepkes sorry for the late reply. We recently change the internal implementation of the HelixAccountService and now there are conflict with your PR. If you are still interested, can you please resolve the conflict so we can merge your change after? Thanks.

**Comment by siepkes** [comment]
Still interested! I'll give you folks a ping when I'm done fixing the merge conflicts.

**Comment by siepkes** [comment]
@justinlin-linkedin I have rebased my changes on `master` and I'm currently in the process of checking everything. However I noticed that `master` currently has tests failures when I try to build it?

When I build the `master` branch (ie. no changes of mine) with `clean allJar allTest` it fails with test errors like:

```
com.github.ambry.frontend.FrontendIntegrationTest > stitchedUploadTest[19] FAILED
    java.lang.IllegalStateException: Should have received response. completion context: ChannelCloseListener::operationComplete
        at com.github.ambry.frontend.FrontendIntegrationTest.getHttpResponse(FrontendIntegrationTest.java:1207)
        at com.github.ambry.frontend.FrontendIntegrationTest.uploadDataChunksAndVerify(FrontendIntegrationTest.java:1248)
        at com.github.ambry.frontend.FrontendIntegrationTest.stitchedUploadTest(FrontendIntegrationTest.java:428)

com.github.ambry.frontend.FrontendIntegrationTest > accountApiTest[19] FAILED
    java.lang.IllegalStateException: Should have received response. completion context: ChannelCloseListener::operationComplete
        at com.github.ambry.frontend.FrontendIntegrationTest.getHttpResponse(FrontendIntegrationTest.java:1207)
        at com.github.ambry.frontend.FrontendIntegrationTest.getAccounts(FrontendIntegrationTest.java:1387)
        at com.github.ambry.frontend.FrontendIntegrationTest.getAccountsAndVerify(FrontendIntegrationTest.java:1362)
        at com.github.ambry.frontend.FrontendIntegrationTest.accountApiTest(FrontendIntegrationTest.java:459)

256 tests completed, 90 failed

> Task :ambry-frontend:test FAILED

FAILURE: Build failed with an exception.

* What went wrong:
Execution failed for task ':ambry-frontend:test'.
```

Is this expected? Should the `allTest` target complete without errors?

**Comment by justinlin-linkedin** [comment]
@siepkes Master branch isn't failing in my side. Is that test failure always happening? or is it just one time thing?

**Comment by siepkes** [comment]
Running with `allTest` always fails for me. I've tested this on my Linux desktop (Fedora 31), my Notebook (CentOS 7) and I've also spun up a SmartOS VM (ie. Solaris derivative) on our Triton cloud which all failed with errors.

On Fedora 31 it dies with:
```
com.github.ambry.frontend.FrontendIntegrationTest > healthCheckRequestTest[19] FAILED
    java.lang.IllegalStateException: Should have received response. completion context: ChannelCloseListener::operationComplete
        at com.github.ambry.frontend.FrontendIntegrationTest.getHttpResponse(FrontendIntegrationTest.java:1207)
        at com.github.ambry.frontend.FrontendIntegrationTest.healthCheckRequestTest(FrontendIntegrationTest.java:281)

com.github.ambry.frontend.FrontendIntegrationTest > optionsTest[19] FAILED
    java.lang.IllegalStateException: Should have received response. completion context: ChannelCloseListener::operationComplete
        at com.github.ambry.frontend.FrontendIntegrationTest.getHttpResponse(FrontendIntegrationTest.java:1207)
        at com.github.ambry.frontend.FrontendIntegrationTest.optionsTest(FrontendIntegrationTest.java:442)

com.github.ambry.frontend.FrontendIntegrationTest > multipartPostGetHeadUpdateDeleteTest[19] FAILED
    java.lang.IllegalStateException: Should have received response. completion context: ChannelCloseListener::operationComplete
        at com.github.ambry.frontend.FrontendIntegrationTest.getHttpResponse(FrontendIntegrationTest.java:1207)
        at com.github.ambry.frontend.FrontendIntegrationTest.verifyPostAndReturnBlobId(FrontendIntegrationTest.java:700)
        at com.github.ambry.frontend.FrontendIntegrationTest.multipartPostBlobAndVerify(FrontendIntegrationTest.java:1171)
        at com.github.ambry.frontend.FrontendIntegrationTest.doPostGetHeadUpdateDeleteTest(FrontendIntegrationTest.java:603)
        at com.github.ambry.frontend.FrontendIntegrationTest.multipartPostGetHeadUpdateDeleteTest(FrontendIntegrationTest.java:249)

com.github.ambry.frontend.FrontendIntegrationTest > stitchedUploadTest[19] FAILED
    java.lang.IllegalStateException: Should have received response. completion context: ChannelCloseListener::operationComplete
        at com.github.ambry.frontend.FrontendIntegrationTest.getHttpResponse(FrontendIntegrationTest.java:1207)
        at com.github.ambry.frontend.FrontendIntegrationTest.uploadDataChunksAndVerify(FrontendIntegrationTest.java:1248)
        at com.github.ambry.frontend.FrontendIntegrationTest.stitchedUploadTest(FrontendIntegrationTest.java:428)

com.github.ambry.frontend.FrontendIntegrationTest > accountApiTest[19] FAILED
    java.lang.IllegalStateException: Should have received response. completion context: ChannelCloseListener::operationComplete
        at com.github.ambry.frontend.FrontendIntegrationTest.getHttpResponse(FrontendIntegrationTest.java:1207)
        at com.github.ambry.frontend.FrontendIntegrationTest.getAccounts(FrontendIntegrationTest.java:1387)
        at com.github.ambry.frontend.FrontendIntegrationTest.getAccountsAndVerify(FrontendIntegrationTest.java:1362)
        at com.github.ambry.frontend.FrontendIntegrationTest.accountApiTest(FrontendIntegrationTest.java:459)

com.github.ambry.frontend.FrontendIntegrationTest > postGetHeadUpdateDeleteTest[19] FAILED
    java.lang.IllegalStateException: Should have received response. completion context: ChannelCloseListener::operationComplete
        at com.github.ambry.frontend.FrontendIntegrationTest.getHttpResponse(FrontendIntegrationTest.java:1207)
        at com.github.ambry.frontend.FrontendIntegrationTest.verifyPostAndReturnBlobId(FrontendIntegrationTest.java:700)
        at com.github.ambry.frontend.FrontendIntegrationTest.postBlobAndVerify(FrontendIntegrationTest.java:691)
        at com.github.ambry.frontend.FrontendIntegrationTest.doPostGetHeadUpdateDeleteTest(FrontendIntegrationTest.java:607)
        at com.github.ambry.frontend.FrontendIntegrationTest.postGetHeadUpdateDeleteTest(FrontendIntegrationTest.java:221)

com.github.ambry.frontend.FrontendIntegrationTest > getClusterMapSnapshotTest[19] FAILED
    java.lang.IllegalStateException: Should have received response. completion context: ChannelCloseListener::operationComplete
        at com.github.ambry.frontend.FrontendIntegrationTest.getHttpResponse(FrontendIntegrationTest.java:1207)
        at com.github.ambry.frontend.FrontendIntegrationTest.getClusterMapSnapshotTest(FrontendIntegrationTest.java:327)

com.github.ambry.frontend.FrontendIntegrationTest > getAndUseSignedUrlTest[19] FAILED
    java.lang.IllegalStateException: Should have received response. completion context: ChannelCloseListener::operationComplete
        at com.github.ambry.frontend.FrontendIntegrationTest.getHttpResponse(FrontendIntegrationTest.java:1207)
        at com.github.ambry.frontend.FrontendIntegrationTest.getAndUseSignedUrlTest(FrontendIntegrationTest.java:375)

256 tests completed, 90 failed

> Task :ambry-frontend:test FAILED

FAILURE: Build failed with an exception.

* What went wrong:
Execution failed for task ':ambry-frontend:test'.
> There were failing tests. See the report at: file:///home/siepkes/git/ambry/ambry-frontend/build/reports/tests/test/index.html

* Try:
Run with --stacktrace option to get the stack trace. Run with --info or --debug option to get more log output. Run with --scan to get full insights.

* Get more help at https://help.gradle.org

BUILD FAILED in 1m 54s
159 actionable tasks: 159 executed

```


On the SmartOS VM where I had never build Ambry before it will eventually die with a `java.net.BindException: Address already in use` however except for SSH there are no services running on that VM so thats probably a timing issue.

```
$ git clone https://github.com/linkedin/ambry.git
Cloning into 'ambry'...
remote: Enumerating objects: 31, done.
remote: Counting objects: 100% (31/31), done.
remote: Compressing objects: 100% (20/20), done.
remote: Total 50297 (delta 4), reused 19 (delta 4), pack-reused 50266
Receiving objects: 100% (50297/50297), 17.03 MiB | 13.87 MiB/s, done.
Resolving deltas: 100% (21971/21971), done.
$ cd ambry/
$ ./gradlew clean allJar allTest
----8<---------------------

com.github.ambry.clustermap.HelixClusterManagerTest > xidTest[5] FAILED
    org.I0Itec.zkclient.exception.ZkException: Unable to start single ZooKeeper server.
        at com.github.ambry.utils.TestUtils$ZkServerWrapper.start(TestUtils.java:267)
        at com.github.ambry.utils.TestUtils$ZkInfo.startZkServer(TestUtils.java:326)
        at com.github.ambry.utils.TestUtils$ZkInfo.<init>(TestUtils.java:319)
        at com.github.ambry.clustermap.HelixClusterManagerTest.<init>(HelixClusterManagerTest.java:141)

        Caused by:
        java.net.BindException: Address already in use
            at sun.nio.ch.Net.bind0(Native Method)
            at sun.nio.ch.Net.bind(Net.java:433)
            at sun.nio.ch.Net.bind(Net.java:425)
            at sun.nio.ch.ServerSocketChannelImpl.bind(ServerSocketChannelImpl.java:223)
            at sun.nio.ch.ServerSocketAdaptor.bind(ServerSocketAdaptor.java:74)
            at sun.nio.ch.ServerSocketAdaptor.bind(ServerSocketAdaptor.java:67)
            at org.apache.zookeeper.server.NIOServerCnxnFactory.configure(NIOServerCnxnFactory.java:90)
            at com.github.ambry.utils.TestUtils$ZkServerWrapper.start(TestUtils.java:264)
            ... 3 more

153 tests completed, 70 failed, 12 skipped

> Task :ambry-clustermap:test FAILED

FAILURE: Build failed with an exception.

* What went wrong:
Execution failed for task ':ambry-clustermap:test'.
> There were failing tests. See the report at: file:///root/ambry/ambry-clustermap/build/reports/tests/test/index.html

* Try:
Run with --stacktrace option to get the stack trace. Run with --info or --debug option to get more log output. Run with --scan to get full insights.

* Get more help at https://help.gradle.org

BUILD FAILED in 13m 7s
157 actionable tasks: 140 executed, 17 up-to-date
```

**Comment by siepkes** [comment]
That's great! Thanks!
